### PR TITLE
Bradfitz/custom server

### PR DESCRIPTION
**Pull request by bradfitz**
Two commits. First is from @half-duplex that was sent in https://github.com/tailscale/tailscale-android/pull/45; the second hides it a bit behind a few taps.

https://user-images.githubusercontent.com/2621/182201806-d7e96844-d935-4574-90a3-50fe2a47822b.mov

### CODE REVIEW

**Comment by half-duplex** [comment]
Thanks!
The pitfalls from my attempt in #45 still apply though, if you want a UX better than "debugging hack":
- The app restart is gross and shouldn't be necessary, but ipnlocal can't update ControlURL after Start() and I couldn't find another clean way to trigger ipnlocal.Start() again. I don't know golang, so don't dare go deeper.
- It should be clear on the login view (and maybe logged-in view) that a custom server is configured, maybe with an x to remove
- It should be more obvious in the "custom server" view how to reset to default (erase the contents of the field and save)

**Comment by fernandoenzo** [comment]
This looks really good! Hope to see it soon in F-Droid!

File: cmd/tailscale/ui.go [hunk 1]
```diff
@@ -271,6 +288,36 @@ func (ui *UI) activeDialog() *bool {
...
+		if now.Sub(hi.at) < recent {
+			filt = append(filt, hi)
+		}
+	}
+	ui.menu.showHistory = filt
+	ui.menu.showHistory = append(ui.menu.showHistory, showChange{
+		at:    now,
+		shown: v,
+	})
+	if len(ui.menu.showHistory) >= 6 {
```
Comment (DentonGentry) [inline cmd/tailscale/ui.go:315]: Ideally we'd clear showHistory somewhere, so it doesn't accumulate forever.

**Review by DentonGentry — APPROVED** [review]

File: cmd/tailscale/ui.go (hunk 1 above)
Comment (bradfitz) [reply cmd/tailscale/ui.go:315]: It's cleaned up on line 310 above. We only ever even allocate a slice backing array large enough to store how many times you can click in 5 seconds.

Comment (bradfitz) [reply cmd/tailscale/ui.go:315]: Clarified this a bit and removed the debug print on line 318

**Comment by DentonGentry** [comment]
This functionality is available in the 1.29.72 in the Open Testing track in the Play Store.
- Join via web: https://play.google.com/apps/testing/com.tailscale.ipn
- Join via Play Store app: https://play.google.com/store/apps/details?id=com.tailscale.ipn

It will be present in the next major release 1.30.0.

**Comment by nnsee** [comment]
Kudos! Thank you for listening to user feedback and reconsidering your stance, and settling on a very reasonable compromise.

**Comment by madjam002** [comment]
It would be amazing if we could see something like this dev menu to set the login server on iOS too!

**Comment by Modificator** [comment]
On version 1.35.0, UI working well, but can not apply custom server url 😅

**Comment by megamorf** [comment]
Just fyi, the headscale support has been broken since 1.34.0 - see https://github.com/tailscale/tailscale/issues/6671

**Comment by tonygauderman** [comment]
I don't see a method to make this work from the Android TV platform.  Once you select the 3 ellipses, I do not see a way to deselect them.  The back arrow, which works in other apps to do the same thing, returns you to the Android TV home screen.

**Comment by frustrate3200** [comment]
@tonygauderman Hi, Have you found a solution for android tv yet? Thank you!

**Comment by basnappl** [comment]
@tonygauderman & @frustrate3200, I suggest trying to use this tool with your Android TV: https://github.com/Genymobile/scrcpy

This might allow you to tap the ellipsis as on a touch-enable android device.

**Comment by godsgood33** [comment]
Running 1.48.2 on my Pixel and it appears to still be broke.  Changed the server, but not allowing me to use the preauthkey.  Anybody found a fix?

**Pull request merged**